File: nim_core_v2/lib/src/friend/friend_service.dart
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

part of nim_core_v2;

/// 好友服务
@HawkEntryPoint()
class FriendService {
  factory FriendService() {
    if (_singleton == null) {
      _singleton = FriendService._();
    }
    return _singleton!;
  }

  FriendService._();

  static FriendService? _singleton;

  FriendServicePlatform get _platform => FriendServicePlatform.instance;

  /// 已添加好友
  /// 触发条件， 本端直接添加好友，多端同步
  @HawkApi(ignore: true)
  Stream<NIMFriend> get onFriendAdded => _platform.onFriendAdded;

  /// 删除好友通知
  /// 触发条件：本端删除好友，多端同步 ， 对方删除你（nimFriendDeletionTypeByFriend）
  @HawkApi(ignore: true)
  Stream<NIMFriendDeletion> get onFriendDeleted => _platform.onFriendDeleted;

  /// 好友信息变更
  /// 触发条件：本端修改好友信息，多端同步
  @HawkApi(ignore: true)
  Stream<NIMFriend> get onFriendInfoChanged => _platform.onFriendInfoChanged;

  /// 好友添加申请回调
  @HawkApi(ignore: true)
  Stream<NIMFriendAddApplication> get onFriendAddApplication =>
      _platform.onFriendAddApplication;

  /// 好友添加申请被拒绝回调
  @HawkApi(ignore: true)
  Stream<NIMFriendAddApplication> get onFriendAddRejected =>
      _platform.onFriendAddRejected;

  /// 添加好友
  /// 调用此方法后，系统会向对端发送一条系统通知。
  /// 通知类型：FRIEND_ADD
  /// 调用此接口后：
  /// - 如果 addMode 是 ADD：本端和对端SDK会触发：onFriendAdded
  /// - 如果 addMode 是 APPLY：对端SDK会触发：onFriendAddApplication
  Future<NIMResult<void>> addFriend(
      String accountId, NIMFriendAddParams? params) async {
    return _platform.addFriend(accountId, params);
  }

  /// 删除好友
  /// 调用此方法后，系统会向对端发送一条系统通知。
  /// 通知类型：FRIEND_DELETE
  /// 调用此接口后，本端和对端SDK都会触发：onFriendDeleted
  Future<NIMResult<void>> deleteFriend(
      String accountId, NIMFriendDeleteParams? params) async {
    return _platform.deleteFriend(accountId, params);
  }

  /// 接受好友申请
  /// 调用此方法后，会向对端发送一条系统通知。
  /// 通知类型：FRIEND_ADD
  /// 调用此接口后，本端SDK会触发：onFriendAdded
  Future<NIMResult<void>> acceptAddApplication(
      NIMFriendAddApplication application) async {
    return _platform.acceptAddApplication(application);
  }

  /// 拒绝添加好友申请
  /// 调用此方法后，会向对端发送一条系统通知。
  /// 通知类型：FRIEND_ADD
  /// 调用此接口后，对端SDK会触发：onFriendAddRejected
  Future<NIMResult<void>> rejectAddApplication(
      NIMFriendAddApplication application, String postscript) async {
    return _platform.rejectAddApplication(application, postscript);
  }

  /// 设置好友信息
  /// 调用此接口后，本端SDK会触发：onFriendsInfoChanged
  Future<NIMResult<void>> setFriendInfo(
      String accountId, NIMFriendSetParams params) async {
    return _platform.setFriendInfo(accountId, params);
  }

  /// 获取好友列表
  /// 本地查询，登录后开始同步好友信息，建议同步完成后拉取一次。
  Future<NIMResult<List<NIMFriend>>> getFriendList() async {
    return _platform.getFriendList();
  }

  /// 根据账号ID获取好友信息
  /// 只返回存在的好友信息，返回顺序与输入顺序一致。
  Future<NIMResult<List<NIMFriend>>> getFriendByIds(
      List<String> accountIds) async {
    return _platform.getFriendByIds(accountIds);
  }

  /// 根据账号ID检查好友状态
  Future<NIMResult<Map<String, bool>>> checkFriend(
      List<String> accountIds) async {
    return _platform.checkFriend(accountIds);
  }

  /// 获取申请添加好友信息列表
  /// 查询verifyType等于3的数据，从新到旧排序。
  Future<NIMResult<NIMFriendAddApplicationResult>> getAddApplicationList(
      NIMFriendAddApplicationQueryOption option) async {
    return _platform.getAddApplicationList(option);
  }

  /// 获取好友申请未读数量
  /// 统计所有未处理且未读的申请数量。
  Future<NIMResult<int>> getAddApplicationUnreadCount() async {
    return _platform.getAddApplicationUnreadCount();
  }

  /// 设置好友申请已读
  /// 调用此方法后，所有历史未读数据将标记为已读。
  Future<NIMResult<void>> setAddApplicationRead() async {
    return _platform.setAddApplicationRead();
  }

  /// 根据关键字搜索好友信息
  Future<NIMResult<List<NIMFriend>>> searchFriendByOption(
      NIMFriendSearchOption friendSearchOption) async {
    return _platform.searchFriendByOption(friendSearchOption);
  }
}

File: nim_core_v2_web/web/src/services/friendService.ts
import { failRes, successRes, emit } from '../utils'
import { NIMResult } from '../types'
import RootService from './rootService'
import { logger } from '../logger'
import { createLoggerDecorator } from '@xkit-yx/utils'
import V2NIM, { V2NIMConst } from 'nim-web-sdk-ng'
import {
  V2NIMCheckFriendResult,
  V2NIMFriend,
  V2NIMFriendAddApplication,
  V2NIMFriendAddApplicationQueryOption,
  V2NIMFriendAddApplicationResult,
  V2NIMFriendAddParams,
  V2NIMFriendDeleteParams,
  V2NIMFriendDeletionType,
  V2NIMFriendSearchOption,
  V2NIMFriendSetParams,
} from 'nim-web-sdk-ng/dist/v2/NIM_BROWSER_SDK/V2NIMFriendService'
import { V2NIMError } from 'nim-web-sdk-ng/dist/v2/NIM_BROWSER_SDK/types'

const TAG_NAME = 'FriendService'
const loggerDec = createLoggerDecorator(TAG_NAME, logger)

class FriendService {
  constructor(private rootService: RootService, private nim: V2NIM) {
    this._onFriendAdded = this._onFriendAdded.bind(this)
    this._onFriendDeleted = this._onFriendDeleted.bind(this)
    this._onFriendAddApplication = this._onFriendAddApplication.bind(this)
    this._onFriendAddRejected = this._onFriendAddRejected.bind(this)
    this._onFriendInfoChanged = this._onFriendInfoChanged.bind(this)

    nim.V2NIMFriendService.on('onFriendAdded', this._onFriendAdded)
    nim.V2NIMFriendService.on('onFriendDeleted', this._onFriendDeleted)
    // 申请添加好友的相关信息，其他端向本端发送好友申请,会触发该事件
    nim.V2NIMFriendService.on(
      'onFriendAddApplication',
      this._onFriendAddApplication
    )
    // 对端拒绝本端好友申请，本端会触发该事件
    nim.V2NIMFriendService.on('onFriendAddRejected', this._onFriendAddRejected)
    // 好友信息更新回调，返回变更的好友信息，包括本端直接更新的好友信息和其他端同步更新的好友信息
    nim.V2NIMFriendService.on('onFriendInfoChanged', this._onFriendInfoChanged)
  }

  destroy(): void {
    this.nim.V2NIMFriendService.off('onFriendAdded', this._onFriendAdded)
    this.nim.V2NIMFriendService.off('onFriendDeleted', this._onFriendDeleted)
    this.nim.V2NIMFriendService.off(
      'onFriendAddApplication',
      this._onFriendAddApplication
    )
    this.nim.V2NIMFriendService.off(
      'onFriendAddRejected',
      this._onFriendAddRejected
    )
    this.nim.V2NIMFriendService.off(
      'onFriendInfoChanged',
      this._onFriendInfoChanged
    )
  }

  @loggerDec
  async addFriend(params: {
    accountId: string
    params?: V2NIMFriendAddParams
  }): Promise<NIMResult<void>> {
    try {
      return successRes(
        await this.nim.V2NIMFriendService.addFriend(
          params.accountId,
          params.params || {
            addMode: V2NIMConst.V2NIMFriendAddMode.V2NIM_FRIEND_MODE_TYPE_ADD,
            postscript: '',
          }
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async deleteFriend(params: {
    accountId: string
    params?: V2NIMFriendDeleteParams
  }): Promise<NIMResult<void>> {
    try {
      return successRes(
        await this.nim.V2NIMFriendService.deleteFriend(
          params.accountId,
          params.params || {
            deleteAlias: true,
          }
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async acceptAddApplication(params: {
    application: V2NIMFriendAddApplication
  }): Promise<NIMResult<void>> {
    try {
      return successRes(
        await this.nim.V2NIMFriendService.acceptAddApplication(
          params.application
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async rejectAddApplication(params: {
    application: V2NIMFriendAddApplication
    postscript: string
  }): Promise<NIMResult<void>> {
    try {
      return successRes(
        await this.nim.V2NIMFriendService.rejectAddApplication(
          params.application,
          params.postscript
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async setFriendInfo(params: {
    accountId: string
    params: V2NIMFriendSetParams
  }): Promise<NIMResult<void>> {
    try {
      return successRes(
        await this.nim.V2NIMFriendService.setFriendInfo(
          params.accountId,
          params.params
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async getFriendList(): Promise<NIMResult<{ friendList: V2NIMFriend[] }>> {
    try {
      return successRes({
        friendList: await this.nim.V2NIMFriendService.getFriendList(),
      })
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async getFriendByIds(params: {
    accountIds: string[]
  }): Promise<NIMResult<{ friendList: V2NIMFriend[] }>> {
    try {
      return successRes({
        friendList: await this.nim.V2NIMFriendService.getFriendByIds(
          params.accountIds
        ),
      })
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async checkFriend(params: {
    accountIds: string[]
  }): Promise<NIMResult<{ result: V2NIMCheckFriendResult }>> {
    try {
      return successRes({
        result: await this.nim.V2NIMFriendService.checkFriend(
          params.accountIds
        ),
      })
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async getAddApplicationList(params: {
    option: V2NIMFriendAddApplicationQueryOption
  }): Promise<NIMResult<V2NIMFriendAddApplicationResult>> {
    try {
      return successRes(
        await this.nim.V2NIMFriendService.getAddApplicationList(params.option)
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async getAddApplicationUnreadCount(): Promise<NIMResult<number>> {
    try {
      return successRes(
        await this.nim.V2NIMFriendService.getAddApplicationUnreadCount()
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async setAddApplicationRead(): Promise<NIMResult<void>> {
    try {
      return successRes(
        await this.nim.V2NIMFriendService.setAddApplicationRead()
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async searchFriendByOption(params: {
    friendSearchOption: V2NIMFriendSearchOption
  }): Promise<NIMResult<{ friendList: V2NIMFriend[] }>> {
    try {
      return successRes({
        friendList: await this.nim.V2NIMFriendService.searchFriendByOption(
          params.friendSearchOption
        ),
      })
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  /**
   * 监听添加好友
   */
  private _onFriendAdded(friend: V2NIMFriend) {
    logger.log('_onFriendAdded', friend)
    emit('FriendService', 'onFriendAdded', friend)
  }

  /**
   * 监听删除好友
   */
  private _onFriendDeleted(
    accountId: string,
    deletionType: V2NIMFriendDeletionType
  ) {
    logger.log('_onFriendDeleted', accountId)
    emit('FriendService', 'onFriendDeleted', { accountId, deletionType })
  }
  /**
   * 申请添加好友的相关信息，其他端向本端发送好友申请,会触发该事件
   */
  private _onFriendAddApplication(application: V2NIMFriendAddApplication) {
    logger.log('_onFriendAddApplication', application)
    emit('FriendService', 'onFriendAddApplication', application)
  }

  /**
   * 对端拒绝本端好友申请，本端会触发该事件
   */
  private _onFriendAddRejected(rejection: V2NIMFriendAddApplication) {
    logger.log('_onFriendAddRejected', rejection)
    emit('FriendService', 'onFriendAddRejected', rejection)
  }

  /**
   * 好友信息更新回调，返回变更的好友信息，包括本端直接更新的好友信息和其他端同步更新的好友信息
   */
  private _onFriendInfoChanged(friend: V2NIMFriend) {
    logger.log('_onFriendInfoChanged', friend)
    emit('FriendService', 'onFriendInfoChanged', friend)
  }
}

export default FriendService
